Hannath T

<h2>CUSTOMER TRANSACTION PREDICTION</h2>

<h5>Goal: </h5>

Identify which customers will make a specific transaction in the future, irrespective of the amount of money transacted.

<h5>Abstract</h5>



In the banking sector, accurately identifying customers who are likely to make future transactions is crucial for effective marketing, customer engagement, and resource optimization. This project focuses on developing a predictive machine learning model to identify such customers based on historical customer data. The dataset was analyzed and preprocessed to address challenges such as class imbalance and varying feature types. Multiple classification algorithms, including Logistic Regression, Decision Tree, Random Forest, Naive Bayes, XGBoost, and HistGradientBoosting, were evaluated using appropriate performance metrics such as ROC-AUC, precision, recall, and accuracy.

<h5>Tasks: </h5>

Task 1:- Prepare a complete data analysis report on the given data.

Task 2:- Create a predictive model which will help the bank to identify which
customer will make transactions in future.

Task 3:- Create a report stating the performance of multiple models on this data and
suggest the best model for production.

TAsk 4:- Create a report which should include challenges you faced on data and
what technique used with proper reason.

**can skip EDA part in this project as feature names are not provided.**

<h5>Attribute information:</h5>

The dataset is anonymized so we cannot know which feature is what.
There are a total of 200 features in this data set along with ID_code
and target columns. The target columns contain 0 and 1 value where 0
means the customer will not do a transaction and 1 means the
customer will do a transaction.

* TP- Model predicted customer will transact and customer did transact
* FP- Model predicted customer will transact and customer didnt transact
* FN- Model predicted customer will NOT transact and customer transacted
* TN- Model predicted customer will NOT transact and customer didnt transact

**datasetv explaination is impossible as no feature names**

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
df=pd.read_csv("train(1).csv")

In [3]:
df

ID_code  target    var_0   var_1    var_2   var_3    var_4  \
0            train_0       0   8.9255 -6.7863  11.9081  5.0930  11.4607   
1            train_1       0  11.5006 -4.1473  13.8588  5.3890  12.3622   
2            train_2       0   8.6093 -2.7457  12.0805  7.8928  10.5825   
3            train_3       0  11.0604 -2.1518   8.9522  7.1957  12.5846   
4            train_4       0   9.8369 -1.4834  12.8746  6.6375  12.2772   
...              ...     ...      ...     ...      ...     ...      ...   
199995  train_199995       0  11.4880 -0.4956   8.2622  3.5142  10.3404   
199996  train_199996       0   4.9149 -2.4484  16.7052  6.6345   8.3096   
199997  train_199997       0  11.2232 -5.0518  10.5127  5.6456   9.3410   
199998  train_199998       0   9.7148 -8.6098  13.6104  5.7930  12.5173   
199999  train_199999       0  10.8762 -5.7105  12.1183  8.0328  11.5577   

          var_5   var_6    var_7  ...  var_190  var_191  var_192  var_193  \
0       -9.2834  5.1187  18.6266  ...   4.4354   3.9642   3.1364   1.6910   
1        7.0433  5.6208  16.5338  ...   7.6421   7.7214   2.5837  10.9516   
2       -9.0837  6.9427  14.6155  ...   2.9057   9.7905   1.6704   1.6858   
3       -1.8361  5.8428  14.9250  ...   4.4666   4.7433   0.7178   1.4214   
4        2.4486  5.9405  19.2514  ...  -1.4905   9.5214  -0.1508   9.1942   
...         ...     ...      ...  ...      ...      ...      ...      ...   
199995  11.6081  5.6709  15.1516  ...   6.1415  13.2305   3.9901   0.9388   
199996 -10.5628  5.8802  21.5940  ...   4.9611   4.6549   0.6998   1.8341   
199997  -5.4086  4.5555  21.5571  ...   4.0651   5.4414   3.1032   4.8793   
199998   0.5339  6.0479  17.0152  ...   2.6840   8.6587   2.7337  11.1178   
199999   0.3488  5.2839  15.2058  ...   8.9842   1.6893   0.1276   0.3766   

        var_194  var_195  var_196  var_197  var_198  var_199  
0       18.5227  -2.3978   7.8784   8.5635  12.7803  -1.0914  
1       15.4305   2.0339   8.1267   8.7889  18.3560   1.9518  
2       21.6042   3.1417  -6.5213   8.2675  14.7222   0.3965  
3       23.0347  -1.2706  -2.9275  10.2922  17.9697  -8.9996  
4       13.2876  -1.5121   3.9267   9.5031  17.9974  -8.8104  
...         ...      ...      ...      ...      ...      ...  
199995  18.0249  -1.7939   2.1661   8.5326  16.6660 -17.8661  
199996  22.2717   1.7337  -2.1651   6.7419  15.9054   0.3388  
199997  23.5311  -1.5736   1.2832   8.7155  13.8329   4.1995  
199998  20.4158  -0.0786   6.7980  10.0342  15.5289 -13.9001  
199999  15.2101  -2.4907  -2.2342   8.1857  12.1284   0.1385  

[200000 rows x 202 columns]

<h5>Basic Checks</h5>

In [4]:
df.shape

(200000, 202)

* 200000 columns and 202 rows
* High dimensionality (200 features)

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 200000 entries, 0 to 199999
Columns: 202 entries, ID_code to var_199
dtypes: float64(200), int64(1), object(1)
memory usage: 308.2+ MB


In [6]:
df.columns

Index(['ID_code', 'target', 'var_0', 'var_1', 'var_2', 'var_3', 'var_4',
       'var_5', 'var_6', 'var_7',
       ...
       'var_190', 'var_191', 'var_192', 'var_193', 'var_194', 'var_195',
       'var_196', 'var_197', 'var_198', 'var_199'],
      dtype='object', length=202)

*  no feature names provided in the dataset name

In [7]:
df.dtypes

ID_code     object
target       int64
var_0      float64
var_1      float64
var_2      float64
            ...   
var_195    float64
var_196    float64
var_197    float64
var_198    float64
var_199    float64
Length: 202, dtype: object

* Datas in float
* Target : 1/0 (int)

In [8]:
df['target'].value_counts()

target
0    179902
1     20098
Name: count, dtype: int64

* class imbalance in data as large number of 0 and less 1 targets

In [9]:
df.isnull().sum().sum()

np.int64(0)

* No null values

In [10]:
df.describe()

target          var_0          var_1          var_2  \
count  200000.000000  200000.000000  200000.000000  200000.000000   
mean        0.100490      10.679914      -1.627622      10.715192   
std         0.300653       3.040051       4.050044       2.640894   
min         0.000000       0.408400     -15.043400       2.117100   
25%         0.000000       8.453850      -4.740025       8.722475   
50%         0.000000      10.524750      -1.608050      10.580000   
75%         0.000000      12.758200       1.358625      12.516700   
max         1.000000      20.315000      10.376800      19.353000   

               var_3          var_4          var_5          var_6  \
count  200000.000000  200000.000000  200000.000000  200000.000000   
mean        6.796529      11.078333      -5.065317       5.408949   
std         2.043319       1.623150       7.863267       0.866607   
min        -0.040200       5.074800     -32.562600       2.347300   
25%         5.254075       9.883175     -11.200350       4.767700   
50%         6.825000      11.108250      -4.833150       5.385100   
75%         8.324100      12.261125       0.924800       6.003000   
max        13.188300      16.671400      17.251600       8.447700   

               var_7          var_8  ...        var_190        var_191  \
count  200000.000000  200000.000000  ...  200000.000000  200000.000000   
mean       16.545850       0.284162  ...       3.234440       7.438408   
std         3.418076       3.332634  ...       4.559922       3.023272   
min         5.349700     -10.505500  ...     -14.093300      -2.691700   
25%        13.943800      -2.317800  ...      -0.058825       5.157400   
50%        16.456800       0.393700  ...       3.203600       7.347750   
75%        19.102900       2.937900  ...       6.406200       9.512525   
max        27.691800      10.151300  ...      18.440900      16.716500   

             var_192        var_193        var_194        var_195  \
count  200000.000000  200000.000000  200000.000000  200000.000000   
mean        1.927839       3.331774      17.993784      -0.142088   
std         1.478423       3.992030       3.135162       1.429372   
min        -3.814500     -11.783400       8.694400      -5.261000   
25%         0.889775       0.584600      15.629800      -1.170700   
50%         1.901300       3.396350      17.957950      -0.172700   
75%         2.949500       6.205800      20.396525       0.829600   
max         8.402400      18.281800      27.928800       4.272900   

             var_196        var_197        var_198        var_199  
count  200000.000000  200000.000000  200000.000000  200000.000000  
mean        2.303335       8.908158      15.870720      -3.326537  
std         5.454369       0.921625       3.010945      10.438015  
min       -14.209600       5.960600       6.299300     -38.852800  
25%        -1.946925       8.252800      13.829700     -11.208475  
50%         2.408900       8.888200      15.934050      -2.819550  
75%         6.556725       9.593300      18.064725       4.836800  
max        18.321500      12.000400      26.079100      28.500700  

[8 rows x 201 columns]

In [11]:
df.duplicated().sum()

np.int64(0)

* no duplicates present in dataset

In [12]:
df.dtypes.value_counts()

float64    200
object       1
int64        1
Name: count, dtype: int64

In [13]:
df['target'].value_counts(normalize=True)

target
0    0.89951
1    0.10049
Name: proportion, dtype: float64

* all the features are float
* target is int 
* only object type is id code

* 200000 rows and 202 columns(large dataset)
  
* No null values present in the dataset

* Target =target(1/0)

* presents of class imbalance (large no:of 0 and less 1)

* Numerical features with no feature names provided

* the dataset is has no feature names ,hence EDA is skipped as given.

<h4>Completed Task 1:- Prepare a complete data analysis report on the given data</h4>

<h5>Data Preprocessing</h5>

separating target

In [14]:
X=df.drop(['target','ID_code'],axis=1)
Y=df['target']

In [15]:
X

var_0   var_1    var_2   var_3    var_4    var_5   var_6    var_7  \
0        8.9255 -6.7863  11.9081  5.0930  11.4607  -9.2834  5.1187  18.6266   
1       11.5006 -4.1473  13.8588  5.3890  12.3622   7.0433  5.6208  16.5338   
2        8.6093 -2.7457  12.0805  7.8928  10.5825  -9.0837  6.9427  14.6155   
3       11.0604 -2.1518   8.9522  7.1957  12.5846  -1.8361  5.8428  14.9250   
4        9.8369 -1.4834  12.8746  6.6375  12.2772   2.4486  5.9405  19.2514   
...         ...     ...      ...     ...      ...      ...     ...      ...   
199995  11.4880 -0.4956   8.2622  3.5142  10.3404  11.6081  5.6709  15.1516   
199996   4.9149 -2.4484  16.7052  6.6345   8.3096 -10.5628  5.8802  21.5940   
199997  11.2232 -5.0518  10.5127  5.6456   9.3410  -5.4086  4.5555  21.5571   
199998   9.7148 -8.6098  13.6104  5.7930  12.5173   0.5339  6.0479  17.0152   
199999  10.8762 -5.7105  12.1183  8.0328  11.5577   0.3488  5.2839  15.2058   

         var_8   var_9  ...  var_190  var_191  var_192  var_193  var_194  \
0      -4.9200  5.7470  ...   4.4354   3.9642   3.1364   1.6910  18.5227   
1       3.1468  8.0851  ...   7.6421   7.7214   2.5837  10.9516  15.4305   
2      -4.9193  5.9525  ...   2.9057   9.7905   1.6704   1.6858  21.6042   
3      -5.8609  8.2450  ...   4.4666   4.7433   0.7178   1.4214  23.0347   
4       6.2654  7.6784  ...  -1.4905   9.5214  -0.1508   9.1942  13.2876   
...        ...     ...  ...      ...      ...      ...      ...      ...   
199995 -0.6209  5.6669  ...   6.1415  13.2305   3.9901   0.9388  18.0249   
199996 -3.6797  6.0019  ...   4.9611   4.6549   0.6998   1.8341  22.2717   
199997  0.1202  6.1629  ...   4.0651   5.4414   3.1032   4.8793  23.5311   
199998 -2.1926  8.7542  ...   2.6840   8.6587   2.7337  11.1178  20.4158   
199999 -0.4541  9.3688  ...   8.9842   1.6893   0.1276   0.3766  15.2101   

        var_195  var_196  var_197  var_198  var_199  
0       -2.3978   7.8784   8.5635  12.7803  -1.0914  
1        2.0339   8.1267   8.7889  18.3560   1.9518  
2        3.1417  -6.5213   8.2675  14.7222   0.3965  
3       -1.2706  -2.9275  10.2922  17.9697  -8.9996  
4       -1.5121   3.9267   9.5031  17.9974  -8.8104  
...         ...      ...      ...      ...      ...  
199995  -1.7939   2.1661   8.5326  16.6660 -17.8661  
199996   1.7337  -2.1651   6.7419  15.9054   0.3388  
199997  -1.5736   1.2832   8.7155  13.8329   4.1995  
199998  -0.0786   6.7980  10.0342  15.5289 -13.9001  
199999  -2.4907  -2.2342   8.1857  12.1284   0.1385  

[200000 rows x 200 columns]

In [16]:
Y

0         0
1         0
2         0
3         0
4         0
         ..
199995    0
199996    0
199997    0
199998    0
199999    0
Name: target, Length: 200000, dtype: int64

<h4>Train Test Split</h4>

In [17]:
from sklearn.model_selection import train_test_split
X_train,X_test,Y_train,Y_test=train_test_split(X,Y,test_size=0.2,stratify=Y,random_state=42)

<h4>scaling</h4>

In [18]:
from sklearn.preprocessing import StandardScaler
scaler=StandardScaler()
X_train_scaled=scaler.fit_transform(X_train)
X_test_scaled=scaler.transform(X_test)

In [19]:
X_train_scaled

array([[ 1.55527004, -0.11525565, -0.09088068, ...,  0.04686222,
        -0.83703878, -1.21708255],
       [ 0.54978094, -1.61339597,  0.40077456, ..., -0.72278067,
        -1.69788529, -0.01119958],
       [-0.41853361, -1.72371245, -1.32179467, ...,  0.59730508,
         0.66648063, -1.02156661],
       ...,
       [ 0.21395995,  0.7198001 ,  1.2994094 , ..., -2.31233371,
        -0.92302719,  0.25785774],
       [-2.37536198,  0.3403865 , -2.40493247, ...,  0.79641991,
        -0.50533128,  0.7542021 ],
       [-0.6892498 ,  0.2017375 ,  0.11776674, ...,  1.41971231,
         1.20268854,  1.60262663]])

In [20]:
X_test_scaled

array([[-0.62858622,  2.08121814, -0.53874339, ..., -1.06325184,
         0.57574643, -0.39108186],
       [-1.03211093, -1.68992093,  1.79280274, ...,  0.39091615,
        -1.2159055 ,  0.44398807],
       [-1.59486111, -1.3049001 , -0.96688246, ..., -0.22879349,
         0.92467816,  1.4210405 ],
       ...,
       [-2.0287834 ,  1.36440801, -1.22291196, ..., -2.36205814,
        -0.0133621 ,  0.10635541],
       [-0.60670916,  1.22069523, -0.88704327, ...,  0.63378413,
        -0.30899496, -0.60425693],
       [-0.48469113,  0.19529043,  0.7796423 , ..., -1.8464658 ,
        -0.49116029, -0.02963525]])

In [21]:
X_train_scaled.shape

(160000, 200)

In [22]:
X_test_scaled.shape

(40000, 200)

<h3>trying different model</h3>

<h4>Model Comparison</h4>

<h4>logistic regression</h4>

Logistic Regression is widely used as a base model and for comparison

In [23]:
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import (roc_auc_score,confusion_matrix,classification_report)
log_model=LogisticRegression(max_iter=1000,
                            class_weight='balanced',
                            random_state=42)
log_model.fit(X_train_scaled,Y_train)

LogisticRegression(class_weight='balanced', max_iter=1000, random_state=42)

In [24]:
log_prob=log_model.predict_proba(X_test_scaled)[:,1]
log_pred = log_model.predict(X_test_scaled)
print(confusion_matrix(Y_test,log_pred))
print(classification_report(Y_test,log_pred))
print("roc_auc:",roc_auc_score(Y_test,log_prob))

[[28219  7761]
 [  903  3117]]
              precision    recall  f1-score   support

           0       0.97      0.78      0.87     35980
           1       0.29      0.78      0.42      4020

    accuracy                           0.78     40000
   macro avg       0.63      0.78      0.64     40000
weighted avg       0.90      0.78      0.82     40000

roc_auc: 0.8598998891036755


* showing good performance

<h4>L2 regularisation</h4>

In [25]:
l2log_model = LogisticRegression(
    penalty='l2',
    C=1.0,
    class_weight='balanced',
    max_iter=1000,
    solver='lbfgs'
)
l2log_model.fit(X_train_scaled, Y_train)

log_prob = l2log_model.predict_proba(X_test_scaled)[:, 1]
log_pred = log_model.predict(X_test_scaled)
print(confusion_matrix(Y_test,log_pred))
print(classification_report(Y_test,log_pred))
print("Logistic Regression (L2) ROC-AUC:", roc_auc_score(Y_test, log_prob))

[[28219  7761]
 [  903  3117]]
              precision    recall  f1-score   support

           0       0.97      0.78      0.87     35980
           1       0.29      0.78      0.42      4020

    accuracy                           0.78     40000
   macro avg       0.63      0.78      0.64     40000
weighted avg       0.90      0.78      0.82     40000

Logistic Regression (L2) ROC-AUC: 0.8598998891036755


from imblearn.over_sampling import SMOTE
smote=SMOTE(random_state=42)
X_train_bal,Y_train_bal=smote.fit_resample(X_train_scaled,Y_train)

* tried smote for balancing but was extreamly slow

<h4>Decision Tree</h4>

In [26]:
from sklearn.tree import DecisionTreeClassifier
dt_model=DecisionTreeClassifier(
    max_depth=6,
    class_weight='balanced',
    random_state=42
)
dt_model.fit(X_train_scaled,Y_train)

DecisionTreeClassifier(class_weight='balanced', max_depth=6, random_state=42)

balancing using class weight and smote effected the tree model,so removed class weight

In [27]:
dt_pred=dt_model.predict(X_test_scaled)
dt_prob=dt_model.predict_proba(X_test_scaled)[:,1]
print(confusion_matrix(Y_test, dt_pred))
print(classification_report(Y_test, dt_pred))
print("ROC-AUC:", roc_auc_score(Y_test, dt_prob))

[[26113  9867]
 [ 2055  1965]]
              precision    recall  f1-score   support

           0       0.93      0.73      0.81     35980
           1       0.17      0.49      0.25      4020

    accuracy                           0.70     40000
   macro avg       0.55      0.61      0.53     40000
weighted avg       0.85      0.70      0.76     40000

ROC-AUC: 0.6412814678691037


very low ROC_AUC value ,checking if any error

In [28]:
print(X_train_scaled.shape)
print(X_test_scaled.shape)

(160000, 200)
(40000, 200)


cheking any (missing values)Nan values

In [29]:
np.isnan(X_train_scaled).sum(), np.isnan(X_test_scaled).sum()

(np.int64(0), np.int64(0))

Random Forest struggles with high-dimensional, large, oversampled datasets due to the computational cost of training multiple deep decision trees.
used min_sample_leaf ,n_jobs,max_depth
still taking time 
so used class weigh balanced instead of smote

<h4>Random Forest Classifier</h4>

In [30]:
from sklearn.ensemble import RandomForestClassifier
rf_model=RandomForestClassifier(
    n_estimators=200,
    max_depth=10,
    class_weight='balanced',
    n_jobs=-1,
    min_samples_leaf=50,
    random_state=42)
rf_model.fit(X_train_scaled,Y_train)

RandomForestClassifier(class_weight='balanced', max_depth=10,
                       min_samples_leaf=50, n_estimators=200, n_jobs=-1,
                       random_state=42)

In [31]:
rf_pred=rf_model.predict(X_test_scaled)
rf_prob=rf_model.predict_proba(X_test_scaled)[:,1]
print(confusion_matrix(Y_test,rf_pred))
print(classification_report(Y_test,rf_pred))
print("ROC_AUC:",roc_auc_score(Y_test,rf_prob))

[[31445  4535]
 [ 1794  2226]]
              precision    recall  f1-score   support

           0       0.95      0.87      0.91     35980
           1       0.33      0.55      0.41      4020

    accuracy                           0.84     40000
   macro avg       0.64      0.71      0.66     40000
weighted avg       0.88      0.84      0.86     40000

ROC_AUC: 0.8135340460012335


* performance of randomforest is not among the best

**GradientBoosting is taking Too much time so going with HistGradientBoosting**

<h4>HistGradientBoosting</h4>

In [32]:
from sklearn.ensemble import HistGradientBoostingClassifier
hgb_model=HistGradientBoostingClassifier(
    max_depth=6,
    learning_rate=0.05,
    max_iter=200,
    random_state=42
)
hgb_model.fit(X_train,Y_train)

HistGradientBoostingClassifier(learning_rate=0.05, max_depth=6, max_iter=200,
                               random_state=42)

In [33]:
hgb_pred=hgb_model.predict(X_test)
hgb_prob=hgb_model.predict_proba(X_test)[:,1]
print(confusion_matrix(Y_test,hgb_pred))
print(classification_report(Y_test,hgb_pred))
print("ROC_AUC:",roc_auc_score(Y_test,hgb_prob))

[[35935    45]
 [ 3711   309]]
              precision    recall  f1-score   support

           0       0.91      1.00      0.95     35980
           1       0.87      0.08      0.14      4020

    accuracy                           0.91     40000
   macro avg       0.89      0.54      0.55     40000
weighted avg       0.90      0.91      0.87     40000

ROC_AUC: 0.8604058155581181


**showing good performance**

<h4>XGBoost</h4>

In [34]:
from xgboost import XGBClassifier
xgb_model=XGBClassifier(
    objective="binary:logistic",
    eval_metric="auc",
    random_state=42)
xgb_model.fit(X_train_scaled,Y_train)


XGBClassifier(base_score=None, booster=None, callbacks=None,
              colsample_bylevel=None, colsample_bynode=None,
              colsample_bytree=None, device=None, early_stopping_rounds=None,
              enable_categorical=False, eval_metric='auc', feature_types=None,
              feature_weights=None, gamma=None, grow_policy=None,
              importance_type=None, interaction_constraints=None,
              learning_rate=None, max_bin=None, max_cat_threshold=None,
              max_cat_to_onehot=None, max_delta_step=None, max_depth=None,
              max_leaves=None, min_child_weight=None, missing=nan,
              monotone_constraints=None, multi_strategy=None, n_estimators=None,
              n_jobs=None, num_parallel_tree=None, ...)

In [35]:
xgb_pred=xgb_model.predict(X_test_scaled)
xgb_prob=xgb_model.predict_proba(X_test_scaled)[:,1]
print(confusion_matrix(Y_test,xgb_pred))
print(classification_report(Y_test,xgb_pred))
print("ROC_AUC:",roc_auc_score(Y_test,xgb_prob))

[[35493   487]
 [ 3052   968]]
              precision    recall  f1-score   support

           0       0.92      0.99      0.95     35980
           1       0.67      0.24      0.35      4020

    accuracy                           0.91     40000
   macro avg       0.79      0.61      0.65     40000
weighted avg       0.90      0.91      0.89     40000

ROC_AUC: 0.8597325040998455


* performance not among the best

<h4>Checking performance of each model</h4>

In [36]:
from sklearn.metrics import accuracy_score, precision_score
results = {
    "Model": [
        "Logistic Regression",
        "Decision Tree",
        "Random Forest",
        "HistGradient Boosting",
        "XGBoost"
    ],
    "ROC-AUC": [
        roc_auc_score(Y_test, log_prob),
        roc_auc_score(Y_test, dt_prob),
        roc_auc_score(Y_test, rf_prob),
        roc_auc_score(Y_test, hgb_prob),
        roc_auc_score(Y_test, xgb_prob)
        
        
    
    ],

"Accuracy": [
        accuracy_score(Y_test, log_pred),
        accuracy_score(Y_test, dt_pred),
        accuracy_score(Y_test, rf_pred),
        accuracy_score(Y_test, hgb_pred),
        accuracy_score(Y_test, xgb_pred)
    
    ],
    
    "Precision": [
        precision_score(Y_test, log_pred),
        precision_score(Y_test, dt_pred),
        precision_score(Y_test, rf_pred),
        precision_score(Y_test, hgb_pred),
        precision_score(Y_test, xgb_pred)
        
    ]
}

results_df = pd.DataFrame(results)
results_df

Model   ROC-AUC  Accuracy  Precision
0    Logistic Regression  0.859900  0.783400   0.286542
1          Decision Tree  0.641281  0.701950   0.166075
2          Random Forest  0.813534  0.841775   0.329241
3  HistGradient Boosting  0.860406  0.906100   0.872881
4                XGBoost  0.859733  0.911525   0.665292

<h4>trying other for comparison</h4>

<h4>KNN</h4>

In [37]:
from sklearn.neighbors import KNeighborsClassifier
knn = KNeighborsClassifier(
    n_neighbors=15,
    weights='distance'
)
knn.fit(X_train_scaled, Y_train)

KNeighborsClassifier(n_neighbors=15, weights='distance')

In [38]:
knn_prob = knn.predict_proba(X_test_scaled)[:, 1]
knn_pred = knn.predict(X_test_scaled)

print("KNN ROC-AUC:", roc_auc_score(Y_test, knn_prob))
print(confusion_matrix(Y_test, knn_pred))
print(classification_report(Y_test, knn_pred))

KNN ROC-AUC: 0.5970579944911352
[[35980     0]
 [ 4020     0]]
              precision    recall  f1-score   support

           0       0.90      1.00      0.95     35980
           1       0.00      0.00      0.00      4020

    accuracy                           0.90     40000
   macro avg       0.45      0.50      0.47     40000
weighted avg       0.81      0.90      0.85     40000



C:\Users\AlaviThorappa\anaconda3\Lib\site-packages\sklearn\metrics\_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
C:\Users\AlaviThorappa\anaconda3\Lib\site-packages\sklearn\metrics\_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
C:\Users\AlaviThorappa\anaconda3\Lib\site-packages\sklearn\metrics\_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))


<h5>NAive Bayes</h5>

In [39]:
from sklearn.naive_bayes import GaussianNB
nb = GaussianNB()
nb.fit(X_train, Y_train)

GaussianNB()

In [40]:
nb_prob = nb.predict_proba(X_test)[:, 1]
nb_pred = nb.predict(X_test)

print("Naive Bayes ROC-AUC:", roc_auc_score(Y_test, nb_prob))
print(confusion_matrix(Y_test, nb_pred))
print(classification_report(Y_test, nb_pred))

Naive Bayes ROC-AUC: 0.8882275393460711
[[35411   569]
 [ 2626  1394]]
              precision    recall  f1-score   support

           0       0.93      0.98      0.96     35980
           1       0.71      0.35      0.47      4020

    accuracy                           0.92     40000
   macro avg       0.82      0.67      0.71     40000
weighted avg       0.91      0.92      0.91     40000



**SVM and GradientBoost is taking too much time ,so skipped both**

**KNN is showing very bad performance,so skipped**

<h4>Compairing all models</h4>

In [41]:
from sklearn.metrics import accuracy_score, precision_score,recall_score
results = {
    "Model": [
        "Logistic Regression",
        "Decision Tree",
        "Random Forest",
        "HistGradient Boosting",
        "Naive bayes",
        "XGBoost"
     
    ],
    "ROC-AUC": [
        roc_auc_score(Y_test, log_prob),
        roc_auc_score(Y_test, dt_prob),
        roc_auc_score(Y_test, rf_prob),
        roc_auc_score(Y_test, hgb_prob),
        roc_auc_score(Y_test, nb_prob),
        roc_auc_score(Y_test, xgb_prob)
       
            ],

"Accuracy": [
        accuracy_score(Y_test, log_pred),
        accuracy_score(Y_test, dt_pred),
        accuracy_score(Y_test, rf_pred),
        accuracy_score(Y_test, hgb_pred),
        accuracy_score(Y_test, nb_pred),
        accuracy_score(Y_test, xgb_pred)
       
      
    
    ],
    
    "Precision": [
        precision_score(Y_test, log_pred),
        precision_score(Y_test, dt_pred),
        precision_score(Y_test, rf_pred),
        precision_score(Y_test, hgb_pred),
        precision_score(Y_test, nb_pred),
        precision_score(Y_test, xgb_pred)
        
    ],
    "Recall": [
        recall_score(Y_test, log_pred),
        recall_score(Y_test, dt_pred),
        recall_score(Y_test, rf_pred),
        recall_score(Y_test, hgb_pred),
        recall_score(Y_test, nb_pred),
        recall_score(Y_test, xgb_pred)
]
}

results_df = pd.DataFrame(results)
results_df

Model   ROC-AUC  Accuracy  Precision    Recall
0    Logistic Regression  0.859900  0.783400   0.286542  0.775373
1          Decision Tree  0.641281  0.701950   0.166075  0.488806
2          Random Forest  0.813534  0.841775   0.329241  0.553731
3  HistGradient Boosting  0.860406  0.906100   0.872881  0.076866
4            Naive bayes  0.888228  0.920125   0.710138  0.346766
5                XGBoost  0.859733  0.911525   0.665292  0.240796

**After compairing six models**

* Decision Tree is worse performing model with lowest values
* XGBoost performance was low compaired to other model with low precision and moderate recall
* RAndom Forest was slow and precision only 0.33
* Logistic Regression showing best recall 
* HistGradient Boosting has the best precision
* Naive bayes has the best ROC_AUC

<h4>BUT</h4>

* HistGradientBoosting is having high precision and low reacall
* Logistic Regression is having high recall but low precision
* Naive bayes having high accuracy but not the best but good precision and moderate recall

**working on threshold tuning to currect this**

<h4>Threshold Tuning on HistGradientBoosting(basemodel)</h4>

* tried with scaled data still got the same result ,so going with X_train and X_test

In [42]:
from sklearn.metrics import precision_recall_curve
hgb_probs = hgb_model.predict_proba(X_test)[:, 1]
precision, recall, thresholds = precision_recall_curve(Y_test, hgb_probs)
best_threshold = thresholds[np.argmax(
    2 * (precision * recall) / (precision + recall)
)]
hgb_pred_custom = (hgb_probs >= best_threshold).astype(int)

In [43]:
print (classification_report(Y_test,hgb_pred_custom))
print(confusion_matrix(Y_test,hgb_pred_custom))
print("ROC_AUC:",roc_auc_score(Y_test, hgb_probs))

              precision    recall  f1-score   support

           0       0.95      0.93      0.94     35980
           1       0.46      0.53      0.49      4020

    accuracy                           0.89     40000
   macro avg       0.70      0.73      0.72     40000
weighted avg       0.90      0.89      0.89     40000

[[33497  2483]
 [ 1900  2120]]
ROC_AUC: 0.8604058155581181


<h4>Threshold tuning on naive bayes</h4>

In [44]:
nb_probs = nb.predict_proba(X_test)[:, 1]
precision, recall, thresholds = precision_recall_curve(Y_test, nb_probs)
f1_scores = 2 * (precision * recall) / (precision + recall + 1e-8)
best_threshold_nb = thresholds[np.argmax(f1_scores)]
nb_pred_tuned = (nb_probs >= best_threshold_nb).astype(int)

In [45]:
print("Naive Bayes Best Threshold:", best_threshold_nb)
print(confusion_matrix(Y_test, nb_pred_tuned))
print(classification_report(Y_test, nb_pred_tuned))

Naive Bayes Best Threshold: 0.22588472906407833
[[33424  2556]
 [ 1544  2476]]
              precision    recall  f1-score   support

           0       0.96      0.93      0.94     35980
           1       0.49      0.62      0.55      4020

    accuracy                           0.90     40000
   macro avg       0.72      0.77      0.74     40000
weighted avg       0.91      0.90      0.90     40000



<h4>Threshold tuning on Logistic regression</h4>

In [46]:
log_probs = log_model.predict_proba(X_test_scaled)[:, 1]
precision, recall, thresholds = precision_recall_curve(Y_test, log_probs)
f1_scores = 2 * (precision * recall) / (precision + recall + 1e-8)
best_threshold_log = thresholds[np.argmax(f1_scores)]
log_pred_tuned = (log_probs >= best_threshold_log).astype(int)

In [47]:
print("Logistic Regression Best Threshold:", best_threshold_log)
print(confusion_matrix(Y_test, log_pred_tuned))
print(classification_report(Y_test, log_pred_tuned))

Logistic Regression Best Threshold: 0.7307121448703158
[[33390  2590]
 [ 1855  2165]]
              precision    recall  f1-score   support

           0       0.95      0.93      0.94     35980
           1       0.46      0.54      0.49      4020

    accuracy                           0.89     40000
   macro avg       0.70      0.73      0.72     40000
weighted avg       0.90      0.89      0.89     40000



<h4>Comparison after tuning</h4>

In [48]:
from sklearn.metrics import accuracy_score, precision_score,recall_score
results = {
    "Model": [
        "Logistic Regression",
        "Decision Tree",
        "Random Forest",
        "HistGradient Boosting",
        "Naive bayes",
        "XGBoost"
     
    ],
    "ROC-AUC": [
        roc_auc_score(Y_test, log_prob),
        roc_auc_score(Y_test, dt_prob),
        roc_auc_score(Y_test, rf_prob),
        roc_auc_score(Y_test, hgb_probs),
        roc_auc_score(Y_test, nb_prob),
        roc_auc_score(Y_test, xgb_prob)
       
            ],

"Accuracy": [
        accuracy_score(Y_test, log_pred_tuned),
        accuracy_score(Y_test, dt_pred),
        accuracy_score(Y_test, rf_pred),
        accuracy_score(Y_test, hgb_pred_custom),
        accuracy_score(Y_test, nb_pred_tuned),
        accuracy_score(Y_test, xgb_pred)
        
    ],
    
    "Precision": [
        precision_score(Y_test, log_pred_tuned),
        precision_score(Y_test, dt_pred),
        precision_score(Y_test, rf_pred),
        precision_score(Y_test, hgb_pred_custom),
        precision_score(Y_test, nb_pred_tuned),
        precision_score(Y_test, xgb_pred)
        
    ],
    "Recall": [
        recall_score(Y_test, log_pred_tuned),
        recall_score(Y_test, dt_pred),
        recall_score(Y_test, rf_pred),
        recall_score(Y_test, hgb_pred_custom),
        recall_score(Y_test, nb_pred_tuned),
        recall_score(Y_test, xgb_pred)
]
}

results_df = pd.DataFrame(results)
results_df

Model   ROC-AUC  Accuracy  Precision    Recall
0    Logistic Regression  0.859900  0.888875   0.455310  0.538557
1          Decision Tree  0.641281  0.701950   0.166075  0.488806
2          Random Forest  0.813534  0.841775   0.329241  0.553731
3  HistGradient Boosting  0.860406  0.890425   0.460569  0.527363
4            Naive bayes  0.888228  0.897500   0.492051  0.615920
5                XGBoost  0.859733  0.911525   0.665292  0.240796

* Naive Bayes have slightly higher score than HistGradient Boosting still not choosing it as final model as Naive Bayes is not widely used for customer targetting and as a final model ,also :
* Assumes feature independence
* Can be unstable under threshold tuning
* Its independence assumptions and probability instability make it less reliable for production.

**HistGradientBoosting is more stable, have best balance of precision and ranking robest to imbalance and industry preferred for tabular data(Recommended model for Production)**

<h4>Completed Task 3:- Create a report stating the performance of multiple models on this data and suggest the best model for production</h4>

**SELECTED HISTGRADIENT BOOSTING AS THE FINAL MODEL**

* ROC_AUC : 0.860
* Accuracy : 0.890
* Precision : 0.460
* Recall : 0.527

<h4>Automated Hperparameter Tuning</h4>

In [49]:
from sklearn.model_selection import RandomizedSearchCV
param_dist = {
    'max_depth': [3, 4, 5, 6, 8],
    'learning_rate': [0.01, 0.03, 0.05, 0.1],
    'max_iter': [100, 200, 300],
    'min_samples_leaf': [20, 50, 100]
}
hgb = HistGradientBoostingClassifier(random_state=42)
random_search = RandomizedSearchCV(
    hgb,
    param_distributions=param_dist,
    scoring='roc_auc',
    n_iter=20,
    cv=3,
    n_jobs=-1,
    random_state=42
)

random_search.fit(X_train, Y_train)
best_hgb = random_search.best_estimator_

<h4>Rerun threshold tuning</h4>

In [50]:
from sklearn.metrics import precision_recall_curve
hgb_probs = best_hgb.predict_proba(X_test)[:, 1]
precision, recall, thresholds = precision_recall_curve(Y_test, hgb_probs)
f1_scores = 2 * (precision * recall) / (precision + recall + 1e-8)
best_threshold = thresholds[np.argmax(f1_scores)]
hgb_pred_custom = (hgb_probs >= best_threshold).astype(int)

In [51]:
print (classification_report(Y_test,hgb_pred_custom))
print(confusion_matrix(Y_test,hgb_pred_custom))
print("ROC_AUC:",roc_auc_score(Y_test, hgb_probs))

              precision    recall  f1-score   support

           0       0.95      0.94      0.95     35980
           1       0.51      0.54      0.53      4020

    accuracy                           0.90     40000
   macro avg       0.73      0.74      0.74     40000
weighted avg       0.90      0.90      0.90     40000

[[33910  2070]
 [ 1842  2178]]
ROC_AUC: 0.8796921520800666


* increase in ROC_AUC and overall performance

<h4>Feature Importance</h4>

Feature Importance analysis was not performed for HistGradientBoostingClassifier as the dataset features are anonymized and the model does not expose feature importance attriburtes

<h4>Permutation Feature Importance</h4>

permutation feature imporatnace work by shuffling one feature at a time,if model performance drop ,feature is imporatant

In [52]:
from sklearn.inspection import permutation_importance
perm_importance = permutation_importance(
    best_hgb,
    X_test,
    Y_test,
    scoring='roc_auc',
    n_repeats=10,
    random_state=42,
    n_jobs=-1
)
feature_importance = pd.DataFrame({
    'Feature': X_test.columns,
    'Importance': perm_importance.importances_mean
}).sort_values(by='Importance', ascending=False)
feature_importance.head(10)

Feature  Importance
81    var_81    0.007895
139  var_139    0.006693
53    var_53    0.006126
12    var_12    0.005622
6      var_6    0.005277
26    var_26    0.004927
110  var_110    0.004927
80    var_80    0.004483
174  var_174    0.004481
146  var_146    0.004427

* var_81 is the feature that is influencing the target the most
* shuffling var_81 affected the model performance the most

<h4>checking for Overfitting or Underfitting</h4>

* Checking ROC_AUC on both train and test
* if Train score approximately equal to test (acceptable model)

In [53]:
train_probs = best_hgb.predict_proba(X_train)[:, 1]
train_auc = roc_auc_score(Y_train, train_probs)
test_probs = best_hgb.predict_proba(X_test)[:, 1]
test_auc = roc_auc_score(Y_test, test_probs)
print("Train ROC-AUC:", train_auc)
print("Test ROC-AUC:", test_auc)

Train ROC-AUC: 0.9633832352443247
Test ROC-AUC: 0.8796921520800666


* Mild overfitting but still acceptable 

<h4>Cross validation</h4>

In [54]:
from sklearn.model_selection import cross_val_score

cv_auc = cross_val_score(
    best_hgb,
    X_train,
    Y_train,
    scoring='roc_auc',
    cv=3,
    n_jobs=-1
)

print("CV ROC-AUC mean:", cv_auc.mean())
print("CV ROC-AUC std:", cv_auc.std())

CV ROC-AUC mean: 0.8779987978358373
CV ROC-AUC std: 0.002538056529942482


* 0.87 ablity to correctly rank transacting vs non transacting customers
* very low std : model performance barely changed accross different folds(model is stable)

**No serious overfitting**

<h4>Completed Task 2:- Create a predictive model which will help the bank to identify which customer will make transactions in future.</h4>

<h4>Challenges Faced  and Technique Used</h4>

1. Class Imbalance :
   * Target variable was imbalance
   * Fewewr customers making transaction
   * This can affect the model training
* Technique used :
   * Tried Smote but failed
   * prioritised ROC_AUC,recall and precision over accuracy
   * Probablity threshold tuning
   * Selected tree based model(naturally handle some imbalance)
2. High False Negative in Initail model :
   * very low recall
   * missed transaction customers
* Technique used :
   * Probablity threshold tuning using precision recall curves
3. Anonymous dataset :
   * No Exploritory data analysis
   * No feature selection
* Technique used :
   * Pwrmutation Feature Importance
4. High recall but low precision and visa versa
   * challenge on model selection
* Technique used :
   * Probablity threshold tuning

<h4>Completed Task 4:- Create a report which should include challenges you faced on data and what technique used with proper reason.</h4>